### PR TITLE
Fixed #31123 -- Added --include-stale-apps option to the remove_stale_contenttypes management command.

### PR DESCRIPTION
**Pull request by gowthamk63**
Ticket: https://code.djangoproject.com/ticket/31123

### CODE REVIEW

File: django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py [hunk 1]
```diff
@@ -14,6 +14,10 @@ def add_arguments(self, parser):
             '--noinput', '--no-input', action='store_false', dest='interactive',
             help='Tells Django to NOT prompt the user for input of any kind.',
         )
+        parser.add_argument(
```
Comment (gowthamk63) [inline django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: I'm not sure about the name for this argument. Please let me know if there's a better name for this
Thanks!

**Comment by kingbuzzman** [comment]
Can you add "Refs #31123" to the beginning of the PR title

File: django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py [hunk 2]
```diff
@@ -14,6 +14,10 @@ def add_arguments(self, parser):
...
+            '--ignore-app-config', action='store_true',
```
Comment (felixxm) [inline django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: `--ignore-app-config` is unclear for me, maybe `--include-stale-apps`.

Comment (felixxm) [reply django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: Please add `default=False`.

File: django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py [hunk 3]
```diff
@@ -14,6 +14,10 @@ def add_arguments(self, parser):
...
+            help="Removes all content types including the ones that are not included in installed apps.",
```
Comment (felixxm) [inline django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: Maybe:
```
Deletes stale content types including ones whose app isn't installed.
```

**Comment by kingbuzzman** [comment]
@felixxm this will be in 3.1 correct?

**Comment by felixxm** [comment]
Yes, if it will be ready on time👍

**Comment by kingbuzzman** [comment]
@felixxm let me know if there is anything we need to fix. Thanks.

File: docs/ref/django-admin.txt [hunk 4]
```diff
@@ -1651,6 +1651,11 @@ the deletion.
 
 Specifies the database to use. Defaults to ``default``.
 
+.. django-admin-option:: --include-stale-apps
+
+Deletes stale content types including ones whose app isn't installed or have
+been deleted. Defaults to ``False``.
+
```
Comment (kingbuzzman) [inline docs/ref/django-admin.txt]: I'm guessing there needs to be a `.. versionadded:: 3.1` somewhere here, but not sure what it should look like.

File: django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py [hunk 5]
```diff
@@ -22,13 +26,18 @@ def add_arguments(self, parser):
     def handle(self, **options):
         db = options['database']
         interactive = options['interactive']
+        include_stale_apps = options['include_stale_apps']
         verbosity = options['verbosity']
 
-        for app_config in apps.get_app_configs():
-            content_types, app_models = get_contenttypes_and_models(app_config, db, ContentType)
+        content_types_by_app = groupby(ContentType.objects.using(db).order_by('app_label', 'model'),
```
Comment (felixxm) [inline django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: We should keep a fast return (see `get_contenttypes_and_models()`) when `router` doesn't allow for migrations and clear cache, e.g.
```python
if not router.allow_migrate_model(db, ContentType):
    return
ContentType.objects.clear_cache()
```

File: django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py [hunk 6]
```diff
@@ -22,13 +26,18 @@ def add_arguments(self, parser):
...
+                                       lambda x: x.app_label)
+
+        for app_label, cts in content_types_by_app:
```
Comment (felixxm) [inline django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: `cts` -> `content_types`

File: django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py (hunk 5 above)
Comment (felixxm) [inline django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: Use hanging indentation.

File: django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py [hunk 7]
```diff
@@ -22,13 +26,18 @@ def add_arguments(self, parser):
...
+            if not include_stale_apps and app_label not in apps.app_configs:
+                continue
+
             to_remove = [
-                ct for (model_name, ct) in content_types.items()
-                if model_name not in app_models
+                ct for ct in cts if ct.model_class() is None
```
Comment (felixxm) [inline django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: I would push minor edits:
```python
        if not router.allow_migrate_model(db, ContentType):
            return
        ContentType.objects.clear_cache()
        
        apps_content_types = itertools.groupby(
            ContentType.objects.using(db).order_by('app_label', 'model'),
            lambda obj: obj.app_label,
        )   
        for app_label, content_types in apps_content_types:
            if not include_stale_apps and app_label not in apps.app_configs:
                continue
            to_remove = [
                ct for ct in content_types if ct.model_class() is None
            ]
            ...
```

File: tests/contenttypes_tests/test_management.py [hunk 8]
```diff
@@ -12,8 +12,6 @@
 
 @modify_settings(INSTALLED_APPS={'append': ['no_models']})
 class RemoveStaleContentTypesTests(TestCase):
-    # Speed up tests by avoiding retrieving ContentTypes for all test apps.
-    available_apps = ['contenttypes_tests', 'no_models', 'django.contrib.contenttypes']
```
Comment (felixxm) [inline tests/contenttypes_tests/test_management.py]: To preserve this optimization you can add and use an `empty_models` app, e.g.
```python
@modify_settings(INSTALLED_APPS={'append': ['no_models', 'empty_models']})
class RemoveStaleContentTypesTests(TestCase):
    # Speed up tests by avoiding retrieving ContentTypes for all test apps.
    available_apps = ['contenttypes_tests', 'no_models', 'empty_models', 'django.contrib.contenttypes']

    ...

    @modify_settings(INSTALLED_APPS={'remove': ['empty_models']}
    def test_contenttypes_removed_for_apps_not_in_installed_apps(self):
        ...
```

File: tests/contenttypes_tests/test_management.py [hunk 9]
```diff
@@ -71,3 +69,10 @@ def test_contenttypes_removed_in_apps_without_models(self):
             call_command('remove_stale_contenttypes', verbosity=2)
         self.assertIn("Deleting stale content type 'no_models | Fake'", stdout.getvalue())
         self.assertEqual(ContentType.objects.count(), self.before_count)
+
+    def test_contenttypes_removed_for_apps_not_in_installed_apps(self):
+        ContentType.objects.create(app_label='stale_app', model='Fake')
+        with mock.patch('builtins.input', return_value='yes'), captured_stdout() as stdout:
+            call_command('remove_stale_contenttypes', include_stale_apps=True, verbosity=2)
+        self.assertIn("Deleting stale content type 'stale_app | Fake'", stdout.getvalue())
+        self.assertEqual(ContentType.objects.count(), self.before_count)
```
Comment (felixxm) [inline tests/contenttypes_tests/test_management.py]: We should check if it works together with deleted models for installed apps. I would add `ContentType.objects.create(app_label='no_models', model='Fake 1')` to this test, e.g.
```python
@modify_settings(INSTALLED_APPS={'remove': ['empty_models']})
def test_contenttypes_removed_for_apps_not_in_installed_apps(self):
    ContentType.objects.create(app_label='no_models', model='Fake 1')
    ContentType.objects.create(app_label='empty_models', model='Fake 2')
    with mock.patch('builtins.input', return_value='yes'), captured_stdout() as stdout:
        call_command('remove_stale_contenttypes', include_stale_apps=True, verbosity=2)
    self.assertIn("Deleting stale content type 'no_models | Fake 1'", stdout.getvalue())
    self.assertIn("Deleting stale content type 'empty_models | Fake 2'", stdout.getvalue())
    self.assertEqual(ContentType.objects.count(), self.before_count)
```

File: docs/ref/django-admin.txt (hunk 4 above)
Comment (felixxm) [reply docs/ref/django-admin.txt]:
```
.. django-admin-option:: --include-stale-apps 

.. versionadded:: 3.1
```

File: tests/contenttypes_tests/test_management.py [hunk 10]
```diff
@@ -13,7 +13,7 @@
 @modify_settings(INSTALLED_APPS={'append': ['no_models']})
```
Comment (felixxm) [inline tests/contenttypes_tests/test_management.py]: We need to add `empty_models` to the `INSTALLED_APPS`:
```
@modify_settings(INSTALLED_APPS={'append': ['no_models', 'empty_models']})
```

File: django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py [hunk 11]
```diff
@@ -1,11 +1,11 @@
+from itertools import groupby
```
Comment (felixxm) [inline django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: `import itertools`

Comment (kingbuzzman) [reply django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: Yup just caught that :P

Comment (kingbuzzman) [reply django/contrib/contenttypes/management/commands/remove_stale_contenttypes.py]: - fixed.

**Comment by kingbuzzman** [comment]
@felixxm thanks for all your help, please let us know of any further changes needed. Thanks.

**Comment by felixxm** [comment]
@gowthamk63 @kingbuzzman Thanks :+1: I pushed final edits.

**Comment by felixxm** [comment]
OK I cannot push them, Can you allow me to push final edits?

**Comment by kingbuzzman** [comment]
I will tell @gowthamk63 to do this as soon as possible, he is asleep at the moment.

**Comment by kingbuzzman** [comment]
He will be another 2 hours, if you want, you can paste the diff and i can apply it so we're not waiting around. I can modify the branch not the PR.

**Comment by felixxm** [comment]
@kingbuzzman np, I opened a new PR with final edits.

Updated in #12588.

**Comment by kingbuzzman** [comment]
@felixxm can i copy over your changes to this PR and re-open this PR, issue is that if you merge that new PR @gowthamk63  wont get a "Contributor" badge.

**Comment by felixxm** [comment]
I'm not sure what a "Contributor" badge means, but if you want I can make him the main author of this change.

**Comment by kingbuzzman** [comment]
@felixxm please.

**Comment by felixxm** [comment]
Done.

**Pull request closed without merging**